### PR TITLE
Cat Nodes API with Protobuf

### PR DESCRIPTION
**Pull request by VachaShah**
### Description
The purpose of this draft PR is to create a new cat nodes API with protobuf as a serialization/de-serialization mechanism for node-to-node communication. We also benchmark the performance results in comparison to the original cat nodes API. This PR creates a separate path for the new API but when these changes are to be merged, a lot of code in the original API can be replaced with the newer approach. For the purpose of this POC, as of now the files are separate.

This PR represents the work done for the POC and is not be merged. The changes will be raised, reviewed and merged in incremental PRs.

### Related Issues
#6844
#1287

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gashutos** [comment]
Strictly speaking from performance dimension,
I was just benchmarkin netty http layer for `search` path. We hardly spend `1 or 2 ms` in netty layer per request. 
Any benchmark available for protobuf ?

**Comment by dblock** [comment]
@VachaShah Given https://github.com/opensearch-project/OpenSearch/issues/6844#issuecomment-1742250229, do you have any ideas about how to bring this into OpenSearch as maybe an experimental feature? What are next steps?

**Comment by VachaShah** [comment]
> @VachaShah Given [#6844 (comment)](https://github.com/opensearch-project/OpenSearch/issues/6844#issuecomment-1742250229), do you have any ideas about how to bring this into OpenSearch as maybe an experimental feature? What are next steps?

I am working on getting the changes from the POC in #9097 to merge in the repo. We are also working on getting the numbers for APIs like search.

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    @FunctionalInterface
+    interface Writer<V> {
+
+        /**
+         * Write {@code V}-type {@code value} to the {@code out}put stream.
+        *
+        * @param out Output to write the {@code value} too
+        * @param value The value to add
+        */
+        void write(OutputStream out, V value) throws IOException;
```
Comment (peternied) [inline libs/core/src/main/java/org/opensearch/core/common/io/stream/ProtobufWriteable.java:45]: Its strange than the writer is via an OutputStream, but the reader isn't a symetrical version that uses InputStream.  Can we align the types to be consistant, be it bytes[] or *Stream?

File: server/src/main/java/org/opensearch/tasks/ProtobufTask.java [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Current protobuf task information
+*
+* @opensearch.internal
+*/
+public class ProtobufTask {
```
Comment (peternied) [inline server/src/main/java/org/opensearch/tasks/ProtobufTask.java:37]: This class is nearly identical to the existing Task, why do we need a largely identical protobuf version? This seems to imply there is coupling between the existing OpenSearch data models and their serialized form.  Can we decompose these relationships more?

**Review by peternied — CHANGES_REQUESTED** [review]
This is really exciting work - thanks for getting this out there.

> For the purpose of this POC, as of now the files are separate.

What our your thoughts on the end state for this change?  I'm resistant to merge so much duplicate code.

**Review by peternied — COMMENTED** [review]
~ Duplicate comment ~

**Comment by VachaShah** [comment]
> This is really exciting work - thanks for getting this out there.
> 
> > For the purpose of this POC, as of now the files are separate.
> 
> What our your thoughts on the end state for this change? I'm resistant to merge so much duplicate code.

Hi @peternied, thank you for your comments! This PR is not be merged, its a draft PR for the work done to get numbers for which I created a parallel API _cat/nodes_protobuf to compare with the original version. The changes to be merged will go in incremental PRs, which I will raise later. This PR is just out here in the draft state for the work done for the POC.

**Comment by VachaShah** [comment]
> > This is really exciting work - thanks for getting this out there.
> > > For the purpose of this POC, as of now the files are separate.
> > 
> > 
> > What our your thoughts on the end state for this change? I'm resistant to merge so much duplicate code.
> 
> Hi @peternied, thank you for your comments! This PR is not be merged, its a draft PR for the work done to get numbers for which I created a parallel API _cat/nodes_protobuf to compare with the original version. The changes to be merged will go in incremental PRs, which I will raise later. This PR is just out here in the draft state for the work done for the POC.

I think I was not clear in the description of the PR, so I added a comment about this as well.

**Comment by dblock** [comment]
Any parts of this PR that _can_ be merged? Otherwise I think we can document this in https://github.com/opensearch-project/OpenSearch/issues/6844 and close it?

**Comment by VachaShah** [comment]
> Any parts of this PR that _can_ be merged? Otherwise I think we can document this in #6844 and close it? (And nice work!)

Thank you @dblock! I will raise incremental PRs for this, so closing this one now. I will make sure to update #6844 to document this.

**Pull request closed without merging**